tools/copyblocks/bucket.go

```diff
@@ -0,0 +1,232 @@
+}
+
+func (bkt *gcsBucket) Copy(ctx context.Context, name string, dstBucket bucket) error {
+	d, ok := dstBucket.(*gcsBucket)
+	if !ok {
+		return errors.New("destination bucket wasn't a gcs bucket")
+	}
+	srcObj := bkt.Object(name)
+	dstObject := d.BucketHandle.Object(name)
+	copier := dstObject.CopierFrom(srcObj)
+	_, err := copier.Run(ctx)
+	return err
+}
+
+func (bkt *gcsBucket) ListPrefix(ctx context.Context, prefix string, recursive bool) ([]string, error) {
+	if len(prefix) > 0 && prefix[len(prefix)-1:] != delim {
+		prefix = prefix + delim
+	}
+
+	q := &storage.Query{
+		Prefix: prefix,
+	}
+	if !recursive {
+		q.Delimiter = delim
+	}
+
+	var result []string
+
+	it := bkt.Objects(ctx, q)
+	for {
+		obj, err := it.Next()
+
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "listPrefix: error listing %v", prefix)
+		}
+
+		path := ""
+		if obj.Prefix != "" { // synthetic directory, only returned when recursive=false
+			path = obj.Prefix
+		} else {
+			path = obj.Name
+		}
+
+		if strings.HasPrefix(path, prefix) {
+			path = strings.TrimPrefix(path, prefix)
+		} else {
+			return nil, errors.Errorf("listPrefix: path has invalid prefix: %v, expected prefix: %v", path, prefix)
+		}
+
+		result = append(result, path)
+	}
+
+	return result, nil
+}
+
+func (bkt *gcsBucket) UploadMarkerFile(ctx context.Context, name string) error {
+	obj := bkt.Object(name)
+	w := obj.NewWriter(ctx)
+	return w.Close()
+}
+
+func (bkt *gcsBucket) Name() string {
+	return bkt.name
+}
+
+type azureBucket struct {
+	azblob.Client
+	containerClient container.Client
+	containerName   string
+}
+
+func (bkt *azureBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	client := bkt.containerClient.NewBlobClient(name)
+	response, err := client.DownloadStream(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	return response.Body, nil
+}
+
+func (bkt *azureBucket) Copy(ctx context.Context, name string, dstBucket bucket) error {
+	sourceClient := bkt.containerClient.NewBlobClient(name)
+	sasURL, err := sourceClient.GetSASURL(sas.BlobPermissions{Read: true}, time.Now(), time.Now().Add(10*time.Minute))
+	if err != nil {
+		return err
+	}
+	d, ok := dstBucket.(*azureBucket)
+	if !ok {
+		return errors.New("destination bucket wasn't a blob storage bucket")
+	}
+	dstClient := d.containerClient.NewBlobClient(name)
+
+	response, err := dstClient.StartCopyFromURL(ctx, sasURL, nil)
+	if err != nil {
+		return err
+	}
+
+	copyStatus := response.CopyStatus
+	backoff := backoff.New(ctx, backoff.Config{
+		MinBackoff: time.Second,
+		MaxBackoff: time.Minute,
+		MaxRetries: 10,
+	})
+
+	for {
+		if copyStatus == nil {
+			return errors.New("no copy status present for blob copy")
+		}
+
+		switch *copyStatus {
+		case blob.CopyStatusTypeSuccess:
+			return nil
+		case blob.CopyStatusTypeFailed:
+			return errors.New("copy failed")
+		case blob.CopyStatusTypeAborted:
+			return errors.New("copy aborted")
```

Comment:
Is there any way to get more detailed information about what went wrong?

Reply:
For the failure case yes, there's a `x-ms-copy-status-description`. Thanks for asking because I hadn't even seen that before. I've added that to the failure error message. For the abort case no, but that's only because that's a result of calling `Abort Copy Blob`, which we never do.

Reply:
Here is the [reference](https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob-properties?tabs=azure-ad) in case you are curious.